.pytool/Plugin/BuildReportEvaluation/BuildReportQueries.py

```diff
@@ -0,0 +1,265 @@
+#
+#  Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+##
+
+from edk2toollib.database.edk2_db import Edk2DB
+from buildreport_table import BrComponent, BrPcd, BrLibrary
+import logging
+
+
+def CompiledButButNotIncludedInAnyFV(edk2db: Edk2DB, reportlevel: str) -> int:
+    """
+    Identifies components that are compiled but not included in any Firmware
+    Volume (FV). This function queries the database to find components that
+    are compiled but not included in any FV, excluding those of type 
+    'APPLICATION' and those whose paths end with 'Lib'. It logs a warning for
+    each such component found and returns the total count of these components.
+    Args:
+        edk2db (Edk2DB): An instance of the Edk2DB class to interact with the database.
+        reportlevel (str): The level of reporting to be used (not utilized in the current implementation).
+    Returns:
+        int: The number of components that are compiled but not included in any FV.
+    """
+    error_count = 0
+    with edk2db.session() as session:
+        results = (
+            session.query(BrComponent)
+            .filter(BrComponent.fv_name == None)
+            .filter(BrComponent.type != "APPLICATION")
+            .filter(BrComponent.path.endswith("Lib") == False)
+            .all()
+        )
+
+        for result in results:
+            logging.warning(f"{result.path} is compiled but not included in any FV")
+            error_count += 1
+    return error_count
+
+def SetupDxeExistsInFV(edk2db: Edk2DB, reportlevel: str) -> int:
+    """
+    Checks if 'SetupDxe' exists in the Firmware Volume (FV) and logs a warning
+    for each occurrence.
+    Args:
+        edk2db (Edk2DB): The database object to query.
+        reportlevel (str): The level of reporting.
+    Returns:
+        int: The number of occurrences of 'SetupDxe' in the FV.
+    """
+    error_count = 0
+    with edk2db.session() as session:
+        results = (
+            session.query(BrComponent)
+            .filter(BrComponent.fv_name != None)
+            .filter(BrComponent.path.contains("SetupDxe"))
+        ).all()
+
+        for result in results:
+            logging.warning(f"{result.path} SetupDxe is being included in FV")
+            error_count += 1
+    return error_count
+
+def PcdConOutUgaSupportIsNotZero(edk2db: Edk2DB, reportlevel: str) -> int:
+    """
+    Checks if the PcdConOutUgaSupport value is not zero in the database and logs a warning for each occurrence.
+    Args:
+        edk2db (Edk2DB): The database connection object.
+        reportlevel (str): The level of the report.
+    Returns:
+        int: The count of occurrences where PcdConOutUgaSupport is not zero.
+    """
+    error_count = 0
+
+    with edk2db.session() as session:
+        results = (
+            session.query(BrComponent)
+            .join(BrPcd)
+            .filter(BrPcd.token == "PcdConOutUgaSupport" and BrPcd.value == "1")

```

Comment:
Similar to driver names leaking globally, I think something needs to be done to constrain the PCD. At least including the PCD token space GUID in addition to the name (which is here).

I think this rolls out into two generic requests: (1) be able to map this PCD to the exact instance it is checking with the full two-part definition (token space GUID and name) (2) do not check for this PCD in builds that are not using the instance the rule was written for.

Reply:
The resulting json files has improved the check to make sure its the gEfiMdeModulePkgTokenSpaceGuid instance that this maps to.

Since Uga has essentially been deprecated since uefi 2.0 when GOP was introduced, this should be acceptable for platforms being compiled today.